File: services/ai/gateway/internal/metrics/collector.go
package metrics

import (
	"sync"
	"sync/atomic"
	"time"
)

// Collector gathers runtime metrics for the Gateway
type Collector struct {
	startTime time.Time

	// Request counters
	totalRequests   atomic.Int64
	successRequests atomic.Int64
	failedRequests  atomic.Int64

	// Security counters
	blockedInjections atomic.Int64

	// Queue stats
	queueEnqueues   atomic.Int64
	queueTimeouts   atomic.Int64
	queueFullErrors atomic.Int64

	// CLI stats
	cliExecutions    atomic.Int64
	cliTimeouts      atomic.Int64
	cliErrors        atomic.Int64
	cliTotalMs       atomic.Int64 // cumulative execution time in ms

	// Tier counters
	mu           sync.RWMutex
	tierRequests map[string]int64

	// Session pruner stats
	sessionsPruned atomic.Int64

	// Usage (free tier)
	usageRejections atomic.Int64
}

// New creates a new metrics collector
func New() *Collector {
	return &Collector{
		startTime:    time.Now(),
		tierRequests: make(map[string]int64),
	}
}

// --- Increment methods ---

func (c *Collector) IncTotalRequests()         { c.totalRequests.Add(1) }
func (c *Collector) IncSuccessRequests()       { c.successRequests.Add(1) }
func (c *Collector) IncFailedRequests()        { c.failedRequests.Add(1) }
func (c *Collector) IncBlockedInjections()     { c.blockedInjections.Add(1) }
func (c *Collector) IncQueueEnqueues()         { c.queueEnqueues.Add(1) }
func (c *Collector) IncQueueTimeouts()         { c.queueTimeouts.Add(1) }
func (c *Collector) IncQueueFullErrors()       { c.queueFullErrors.Add(1) }
func (c *Collector) IncCLIExecutions()         { c.cliExecutions.Add(1) }
func (c *Collector) IncCLITimeouts()           { c.cliTimeouts.Add(1) }
func (c *Collector) IncCLIErrors()             { c.cliErrors.Add(1) }
func (c *Collector) IncSessionsPruned(n int64) { c.sessionsPruned.Add(n) }
func (c *Collector) IncUsageRejections()       { c.usageRejections.Add(1) }

func (c *Collector) AddCLIDuration(ms int64) {
	c.cliTotalMs.Add(ms)
}

func (c *Collector) IncTierRequest(tier string) {
	c.mu.Lock()
	c.tierRequests[tier]++
	c.mu.Unlock()
}

// --- Snapshot ---

// Snapshot returns a point-in-time snapshot of all metrics
type Snapshot struct {
	UptimeSeconds     float64          `json:"uptime_seconds"`
	TotalRequests     int64            `json:"total_requests"`
	SuccessRequests   int64            `json:"success_requests"`
	FailedRequests    int64            `json:"failed_requests"`
	BlockedInjections int64            `json:"blocked_injections"`
	QueueEnqueues     int64            `json:"queue_enqueues"`
	QueueTimeouts     int64            `json:"queue_timeouts"`
	QueueFullErrors   int64            `json:"queue_full_errors"`
	CLIExecutions     int64            `json:"cli_executions"`
	CLITimeouts       int64            `json:"cli_timeouts"`
	CLIErrors         int64            `json:"cli_errors"`
	CLIAvgMs          float64          `json:"cli_avg_ms"`
	SessionsPruned    int64            `json:"sessions_pruned"`
	UsageRejections   int64            `json:"usage_rejections"`
	RequestsByTier    map[string]int64 `json:"requests_by_tier"`
}

func (c *Collector) Snapshot() Snapshot {
	execs := c.cliExecutions.Load()
	var avgMs float64
	if execs > 0 {
		avgMs = float64(c.cliTotalMs.Load()) / float64(execs)
	}

	c.mu.RLock()
	tierCopy := make(map[string]int64, len(c.tierRequests))
	for k, v := range c.tierRequests {
		tierCopy[k] = v
	}
	c.mu.RUnlock()

	return Snapshot{
		UptimeSeconds:     time.Since(c.startTime).Seconds(),
		TotalRequests:     c.totalRequests.Load(),
		SuccessRequests:   c.successRequests.Load(),
		FailedRequests:    c.failedRequests.Load(),
		BlockedInjections: c.blockedInjections.Load(),
		QueueEnqueues:     c.queueEnqueues.Load(),
		QueueTimeouts:     c.queueTimeouts.Load(),
		QueueFullErrors:   c.queueFullErrors.Load(),
		CLIExecutions:     execs,
		CLITimeouts:       c.cliTimeouts.Load(),
		CLIErrors:         c.cliErrors.Load(),
		CLIAvgMs:          avgMs,
		SessionsPruned:    c.sessionsPruned.Load(),
		UsageRejections:   c.usageRejections.Load(),
		RequestsByTier:    tierCopy,
	}
}

File: services/ai/gateway-2.0/src/core/metrics/collector.ts
/**
 * Metrics collector – ported from Go.
 *
 * Node.js is single-threaded so we don't need atomic operations or mutexes;
 * plain numeric fields and object properties are sufficient.
 */

export interface MetricsSnapshot {
  uptimeSeconds: number;
  totalRequests: number;
  successRequests: number;
  failedRequests: number;
  blockedInjections: number;
  queueEnqueues: number;
  queueTimeouts: number;
  queueFullErrors: number;
  cliExecutions: number;
  cliTimeouts: number;
  cliErrors: number;
  cliAvgMs: number;
  sessionsPruned: number;
  usageRejections: number;
  requestsByTier: Record<string, number>;
}

export class MetricsCollector {
  private readonly startTime = Date.now();
  private totalRequests = 0;
  private successRequests = 0;
  private failedRequests = 0;
  private blockedInjections = 0;
  private queueEnqueues = 0;
  private queueTimeouts = 0;
  private queueFullErrors = 0;
  private cliExecutions = 0;
  private cliTimeouts = 0;
  private cliErrors = 0;
  private cliTotalMs = 0;
  private sessionsPruned = 0;
  private usageRejections = 0;
  private readonly requestsByTier: Record<string, number> = {};

  // Increment methods
  incTotalRequests(): void { this.totalRequests++; }
  incSuccessRequests(): void { this.successRequests++; }
  incFailedRequests(): void { this.failedRequests++; }
  incBlockedInjections(): void { this.blockedInjections++; }
  incQueueEnqueues(): void { this.queueEnqueues++; }
  incQueueTimeouts(): void { this.queueTimeouts++; }
  incQueueFullErrors(): void { this.queueFullErrors++; }
  incCLIExecutions(): void { this.cliExecutions++; }
  incCLITimeouts(): void { this.cliTimeouts++; }
  incCLIErrors(): void { this.cliErrors++; }
  incSessionsPruned(n: number): void { this.sessionsPruned += n; }
  incUsageRejections(): void { this.usageRejections++; }
  addCLIDuration(ms: number): void { this.cliTotalMs += ms; }
  incTierRequest(tier: string): void { this.requestsByTier[tier] = (this.requestsByTier[tier] ?? 0) + 1; }

  snapshot(): MetricsSnapshot {
    const avgMs = this.cliExecutions > 0 ? this.cliTotalMs / this.cliExecutions : 0;
    return {
      uptimeSeconds: (Date.now() - this.startTime) / 1000,
      totalRequests: this.totalRequests,
      successRequests: this.successRequests,
      failedRequests: this.failedRequests,
      blockedInjections: this.blockedInjections,
      queueEnqueues: this.queueEnqueues,
      queueTimeouts: this.queueTimeouts,
      queueFullErrors: this.queueFullErrors,
      cliExecutions: this.cliExecutions,
      cliTimeouts: this.cliTimeouts,
      cliErrors: this.cliErrors,
      cliAvgMs: Math.round(avgMs * 100) / 100,
      sessionsPruned: this.sessionsPruned,
      usageRejections: this.usageRejections,
      requestsByTier: { ...this.requestsByTier },
    };
  }
}
